File: hackerrank/problem solving/algorithms/2 - implementation/4 - between two sets/with morfo library/between_two_sets.cs
// https://www.hackerrank.com/challenges/between-two-sets/problem?isFullScreen=true

using Morfo.IO;
using Morfo.Math;

Reader.SkipInputLn();
int[] a = Reader.ReadLn<int, int[]>();
int[] b = Reader.ReadLn<int, int[]>();
Console.WriteLine(BetweenTwoSets(a, b));

// n1, n2: length of a, b
// k: number of multiples tested = gcdOfB / lcmOfA
// T: O(n1 + n2 + k)
// S: O(1) extra space
int BetweenTwoSets(int[] a, int[] b)
{
    int lcmOfA = a.Lcm();
    int gcdOfB = b.Gcd();

    int count = 0;
    for (int i = lcmOfA; i <= gcdOfB; i += lcmOfA)
        if (gcdOfB % i == 0)
            count++;
    return count;
}

File: hackerrank/problem solving/algorithms/2 - implementation/4 - between two sets/with morfo library/between_two_sets.java
// https://www.hackerrank.com/challenges/between-two-sets/problem?isFullScreen=true
// Java 25

import morfo.io.Reader;
import morfo.math.Gcd;
import morfo.math.Lcm;

void main() {
    Reader.skipInputLn();
    int[] a = Reader.readLnInts();
    int[] b = Reader.readLnInts();
    IO.println(betweenTwoSets(a, b));
}

// n1, n2: length of a, b
// k: number of multiples tested = gcdOfB / lcmOfA
// T: O(n1 + n2 + k)
// S: O(1) extra space
int betweenTwoSets(int[] a, int[] b) {
    int lcmOfA = Lcm.ofInts(a);
    int gcdOfB = Gcd.ofInts(b);
    return (int) IntStream.iterate(lcmOfA, i -> i <= gcdOfB, i -> i + lcmOfA)
            .filter(i -> gcdOfB % i == 0)
            .count();
}
